sys/align.go

```diff
@@ -31,6 +36,88 @@ func initAlign() {
 	}
 }
 
+func markBitfieldLast(t *Type) {
+	switch t1 := (*t).(type) {
+	case *IntType:
+		t1.BitfieldLst = true
+	case *ConstType:
+		t1.BitfieldLst = true
+	case *LenType:
+		t1.BitfieldLst = true
+	case *FlagsType:
+		t1.BitfieldLst = true
+	case *ProcType:
+		t1.BitfieldLst = true
+	default:
+		panic(fmt.Sprintf("type %+v can't be a bitfield", t1))
+	}
+}
+
+func setBitfieldOffset(t *Type, offset uintptr) {
+	switch t1 := (*t).(type) {
+	case *IntType:
+		t1.BitfieldOff = offset
+	case *ConstType:
+		t1.BitfieldOff = offset
+	case *LenType:
+		t1.BitfieldOff = offset
+	case *FlagsType:
+		t1.BitfieldOff = offset
+	case *ProcType:
+		t1.BitfieldOff = offset
+	default:
+		panic(fmt.Sprintf("type %+v can't be a bitfield", t1))
+	}
+}
+
+func markBitfields(t *StructType) {
+	var bfTypeSize, bfOffset uintptr
+	for i, f := range t.Fields {
+		bfLength := f.BitfieldLength()
+		if bfLength == 0 {
+			// If the current field is not a bitfield,
+			if bfOffset != 0 {
+				// but the previous one is.
+				markBitfieldLast(&t.Fields[i-1])
+				bfTypeSize = 0
+				bfOffset = 0
+			}
+		} else {
+			// If the current field is a bitfield,
+			size := f.Size()
+			if bfOffset != 0 {
+				// and the previous one is as well.
+				if bfTypeSize != size {
+					// The previous bitfield has base type of different size.
+					markBitfieldLast(&t.Fields[i-1])
+					bfOffset = 0
+					bfTypeSize = size
+				}
+				if bfOffset+bfLength > size*8 {
+					// The new bitfield doesn't fit into the base type after the previous ones.
+					if bfOffset == 0 {
+						panic(fmt.Sprintf("bitfield of size %v is too large for base type of size %v", bfLength, size))
+					}
+					markBitfieldLast(&t.Fields[i-1])
+					bfOffset = 0
+				}
+				setBitfieldOffset(&f, bfOffset)
+				bfOffset += bfLength
+			} else {
+				// but the previous one isn't.
+				bfTypeSize = size
+				bfOffset = bfLength
+			}
+		}
+	}
+	if bfOffset != 0 {
+		// The last field in the struct is a bitfield.
+		markBitfieldLast(&t.Fields[len(t.Fields)-1])
```

Comment:
You call markBitfieldLast from 4 different places, it does not look right. I think this function can be more concise, something along the lines of:  

```go
func markBitfields(t *StructType) {
	var bfOffset uintptr
	for i, f := range t.Fields {
		if f.BitfieldLength() == 0 {
			continue
		}
		off0, last := bfOffset, false
		bfOffset += f.BitfieldLength()
		if i == len(t.Fields) - 1 || // last bitfield in a group, if last field of the struct...
			t.Fields[i+1].BitfieldLength() == 0 || // or next field is not a bitfield...
			f.Size() != t.Fields[i+1].Size() || // or next field is of different size...
			bfOffset + t.Fields[i+1].BitfieldLength() > f.Size() * 8 { // of next field does not fit into the current group
			last, bfOffset = true, 0
		}
		setBitfieldOffset(f, last, off)
	}
}
```

This will also get rid of duplication in markBitfieldLast/setBitfieldOffset.

The `bitfield of size %v is too large for base type of size %v` check should go into sysgen/parseIntType.